proxy/src/connection.rs

```diff
@@ -152,20 +169,21 @@ impl Future for Connecting {
     type Error = io::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let socket = try_ready!(self.0.poll());
+        let socket = try_ready!(self.inner.poll());
         set_nodelay_or_warn(&socket);
-        Ok(Async::Ready(Connection::new(Box::new(socket))))
+        Ok(Async::Ready(Connection::new(Box::new(socket), self.is_tls)))
     }
 }
 
 // ===== impl Connection =====
 
 impl Connection {
     /// A constructor of `Connection` with a plain text TCP socket.
-    fn new(io: Box<Io>) -> Self {
+    fn new(io: Box<Io>, is_tls: bool) -> Self {
```

Comment:
i'd be inclined to have a different constructor for each, rather than pass a flag

Reply:
also, i'd probably recommend making the boxing happen within the constructor to reduce callsite complexity

Reply:
I thought about that, but eventually the `TlsStatus` enum that gets passed in here as of 5d6a9fb85f808eef3e0998605554ff2bc2a1362d will have at least three variants, and possibly more ("TLS handshake succeeded", "TLS is disabled", "TLS is enabled but we don't have valid configs yet"), so I thought that would lead to a whole lot of different constructors eventually.

Reply:
> also, i'd probably recommend making the boxing happen within the constructor to reduce callsite complexity

I considered that as well, but didn't do it as it seemed like an unrelated change. With your endorsement I'll definitely change that :)

Reply:
my main concern is that the calls to `new` are getting a bit heavyweight. Things like the following are especially hard to parse visually:
```
(Connection::new(Box::new(socket), TlsStatus::Disabled), remote_addr)
```

This can be alleviated somewhat by splitting these lines out into multiple statements, though I might opt just to do push the boxing down... (esp if we're never passing an already-boxed type into new, which seems to be the case).

Reply:
Hopefully fa36a7e64e69d32241e454f084fe081601d34d24 makes this look a little bit nicer?